services/delegation/src/modules/delegation/delegation.service.ts

```diff
@@ -590,4 +604,69 @@ export class DelegationService {
       .getRawOne();
     return parseInt(d.total, 10) || 0;
   }
+
+  private async undelegateFromChain(
+    nodeId: number,
+    type: string,
+    address: string,
+    amount: BigNumber,
+  ) {
+    let undelegationBlockNumber;
+    const targetChain =
+      type === NodeType.MAINNET ? NodeType.MAINNET : NodeType.TESTNET;
+    if (type === NodeType.TESTNET) {
+      undelegationBlockNumber = await this.testnetBlockchainService.undelegate(
+        address,
+        amount,
+      );
+    }
+
+    if (type === NodeType.MAINNET) {
+      undelegationBlockNumber = await this.mainnetBlockchainService.undelegate(
+        address,
+        amount,
+      );
+    }
+    if (undelegationBlockNumber) {
+      const node = await this.nodeRepository.findOne({ id: nodeId });
+      const undelegation = this.undelegationRepository.create();
+      undelegation.address = address;
+      undelegation.value = amount.toString();
+      undelegation.creationBlock = undelegationBlockNumber;
+      undelegation.createdAt = new Date();
+      undelegation.node = node;
+      undelegation.chain = targetChain;
+      undelegation.confirmationBlock =
+        undelegationBlockNumber + this.undelegationBlockDelay;
+      const savedUndelegation = await this.undelegationRepository.save(
+        undelegation,
+      );
+      if (!savedUndelegation) {
+        throw new InternalServerErrorException(
+          `Could not save undelegation for ${address} with value ${amount} from node ${nodeId} on ${type}`,
+        );
+      }
+    }
+  }
+
+  public async confirmUndelegation(undelegation: Undelegation) {
+    if (undelegation.address) {
```

Comment:
When is it not set?

Reply:
Should always be, its just a sanity check because contract calls will fail in case it isn't